File: python/3381.py
from typing import List

class Solution:
    def maxSubarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        min_prefix = [float('inf')] * k
        min_prefix[0] = 0
        prefix = 0
        result = -float('inf')
        for i in range(n):
            prefix += nums[i]
            remainder = (i + 1) % k
            if min_prefix[remainder] != float('inf'):
                result = max(result, prefix - min_prefix[remainder])
            min_prefix[remainder] = min(min_prefix[remainder], prefix)
        return result

File: cpp/3381.cpp
class Solution {
public:
    long long maxSubarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        
        // Track minimum prefix sum for each length remainder mod k
        // minPrefix[r] = minimum prefix sum seen so far at position % k == r
        vector<long long> minPrefix(k, LLONG_MAX);
        minPrefix[0] = 0; // prefix[0] = 0, and 0 % k == 0
        
        long long prefix = 0; // Current prefix sum
        long long result = LLONG_MIN;
        
        for (int i = 0; i < n; i++) {
            prefix += nums[i]; // prefix[i+1]
            int remainder = (i + 1) % k;
            
            // If we have seen a prefix with the same remainder, we can form a valid subarray
            // Sum = prefix - minPrefix[remainder]
            // Length = (i+1) - start_position, which is divisible by k
            if (minPrefix[remainder] != LLONG_MAX)
                result = max(result, prefix - minPrefix[remainder]);
            
            // Update minimum prefix sum for this remainder
            minPrefix[remainder] = min(minPrefix[remainder], prefix);
        }
        
        return result;
    }
};